torch/onnx/_type_utils.py

```diff
@@ -108,36 +114,102 @@ def from_name(
+        """Convert a torch dtype to JitScalarType.
+
+        Note: DO NOT USE this API when `dtype` comes from a `torch._C.Value.type()` calls.
+            A "RuntimeError: INTERNAL ASSERT FAILED at "../aten/src/ATen/core/jit_type_base.h" can
+            be raised in several scenarios where shape info is not present.
+            Instead use `from_value` API which is safer.
+
+        Args:
+            dtype: A torch.dtype to create a JitScalarType from
+
+        Returns:
+            JitScalarType
+
+        Raises:
+            ValueError: if dtype is not a valid torch.dtype or if it is None.
+        """
         if dtype not in _DTYPE_TO_SCALAR_TYPE:
             raise ValueError(f"Unknown dtype: {dtype}")
         return _DTYPE_TO_SCALAR_TYPE[dtype]
 
+    # TODO: `value: torch._C.Value` type annotation raises (CI machines only):
+    #       NameError: name '_C' is not defined
+    # In both cases, using typing.TypeVar or string literals didn't work
+    @classmethod
+    @_beartype.beartype
+    def from_value(cls, value, default=None) -> JitScalarType:
+        """Create a JitScalarType from a torch._C.Value's underlying scalar type.
+
+        Args:
+            value: A `Union[torch._C.Value, torch.Tensor]` object to fetch scalar type from.
+            default: The JitScalarType to return if a valid scalar cannot be fetched from value
+
+        Returns:
+            JitScalarType.
+
+        Raises:
+            OnnxExporterError: if value does not have a valid scalar type and default is not None.
+            SymbolicValueError: when value.type()'s info are empty and default is None
+        """
+
+        if value is None or not isinstance(value, (torch._C.Value, torch.Tensor)):
+            # default value takes precedence when value is not valid
+            if default is None:
+                raise errors.OnnxExporterError(
+                    "value must be either torch._C.Value, torch.Tensor objects or None."
+                )
+            elif not isinstance(default, JitScalarType):
+                raise errors.OnnxExporterError(
+                    "default value must be a JitScalarType object or None."
+                )
+            return default
+
+        # Each value type has their own way of storing scalar type
+        if isinstance(value, torch.Tensor):
+            return cls.from_dtype(value.dtype)
+        if isinstance(value.type(), torch.ListType):
+            return cls.from_dtype(value.type().getElementType().dtype())
+
+        # isinstance(value, torch._C.Value) is a scalar (non-list)
+        scalar_type = value.type().scalarType()
+        if scalar_type is not None:
+            return cls.from_name(scalar_type)
+
+        # When everything fails... try to default
+        if default is not None:
+            return default
+        raise errors.SymbolicValueError(
+            f"Cannot determine scalar type for this '{type(value.type())}' instance.",
```

Comment:
Do we want to provide a recommendation on what to do?

Reply:
There is no recommendation possible. The user has used the API correctly and `value` is evaluated in runtime, only. There is no way they could prevent this

Reply:
SG